### PR TITLE
Syntactic Sugar! Dag inference, operator composition, and a big docs update

### PR DESCRIPTION
**Pull request by jlowin**
Lots of sugar in this PR, knocking out a whole bunch of roadmap items described [here](https://github.com/airbnb/airflow/wiki/Proposal:-streamlined-workflow). The idea is to make writing DAGfiles easier and less verbose, particularly where code gets extremely repetitive.

Also there is a big update to the "Concepts" doc, which was getting stale, including tips for writing airflow code.

The new features are well summarized (with examples!) in the new docs page, but briefly:
1. DAG assignment can be deferred until after an Operator is created
2. DAG assignment can be inferred from setting a relationship with an Operator that already has a DAG assigned
3. DAGs can be used as context managers to assign Operators 
4. Operators and DAGs can be composed (setting upstream/downstream/assignment relationships) with the `>>` and `<<` bitshift operators
5. Operators have a default owner: `Airflow` (a minor thing, but assigning owners over and over was annoying. In theory users can make a `policy` to require a different owner if they choose.)

In short, to show it all off, this is now a valid (if extreme) Airflow DAGfile:

``` python
from __future__ import print_function
from airflow import DAG
from airflow.operators import DummyOperator, BashOperator, PythonOperator
from datetime import datetime

# this context manager is superfluous because the dag 
# is also part of the composition pipeline
#
# also note that context manager objects enter globals() 
# as long as the "as" statement is present, so this DAG 
# would be loaded by airflow
with DAG('my_dag', start_date=datetime(2016, 1, 1)) as dag:
    (
        dag # alternatively, this line is unnecessary because of the context manager
        >> DummyOperator(task_id='dummy_1')
        >> BashOperator(
            task_id='bash_1',
            bash_command='echo "HELLO!"')
        >> PythonOperator(
            task_id='python_1',
            python_callable=lambda: print("GOODBYE!"))
    )

```

NB: Had to remove one unit test (the infamous "trap executor error" that has caused some confusion) -- it relied on being able to raise a very unusual error and deferred dag assignment "fixed" the problem I was taking advantage of. I can't find another stable way to break the system...

### CODE REVIEW

**Comment by mistercrunch** [comment]
PR OF THE WEEK AWARD!
img

**Comment by bolkedebruin** [comment]
+1 !!

**Comment by syvineckruyk** [comment]
:+1:

File: airflow/configuration.py [hunk 1]
```diff
@@ -93,6 +93,9 @@ def run_command(command):
         'sql_alchemy_pool_recycle': 3600,
         'dagbag_import_timeout': 30,
     },
+    'operators': {
+        'default_owner': 'Airflow'
```
Comment (criccomini) [inline airflow/configuration.py]: Since this is supposed to be a unix-y thing, perhaps `airflow` instead of `Airflow`?

File: airflow/models.py [hunk 2]
```diff
@@ -3057,8 +3190,8 @@ def delete(cls, xcoms, session=None):
             xcoms = [xcoms]
         for xcom in xcoms:
             if not isinstance(xcom, XCom):
-                raise TypeError(
-                    'Expected XCom; received {}'.format(type(xcom)))
+                raise TypeError('Expected XCom; received {}'.format(
+                    xcom.__class__.__name__))
```
Comment (criccomini) [inline airflow/models.py]: Nit: PEP8 enforces alignment like this, I think:

```
            raise TypeError('Expected XCom; received {}'.format(
                            xcom.__class__.__name__))
```

File: tests/dags/test_raise_executor_error.py [hunk 3]
```diff
@@ -1,49 +0,0 @@
-# -*- coding: utf-8 -*-
```
Comment (criccomini) [inline tests/dags/test_raise_executor_error.py:1]: Why is this deleted?

**Comment by criccomini** [comment]
I'm +1.

Left a few comments/questions. Make sure to squash and comment properly before merging.
